src/lib/homestore/shard_manager.cpp

```diff
@@ -61,183 +62,119 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id pg_owner, ui
+        (homestore::ReplicationService*)(&homestore::HomeStore::instance()->repl_service());
+    auto repl_dev = replication_service->get_replica_dev(pg.value().repl_dev_uuid);
+    if (!bool(repl_dev)) {
         LOGWARN("failed to get replica set instance for pg [{}]", pg_owner);
         return folly::makeUnexpected(ShardError::UNKNOWN_PG);
     }
 
-    auto replica_set = std::get< home_replication::rs_ptr_t >(replica_set_var);
     auto new_shard_id = generate_new_shard_id(pg_owner);
     auto create_time = get_current_timestamp();
     auto shard_info =
         ShardInfo(new_shard_id, pg_owner, ShardInfo::State::OPEN, create_time, create_time, size_bytes, size_bytes, 0);
     std::string create_shard_message = serialize_shard(Shard(shard_info));
-    // preapre msg header;
-    const uint32_t needed_size = sizeof(ReplicationMessageHeader) + create_shard_message.size();
-    auto buf = nuraft::buffer::alloc(needed_size);
-    uint8_t* raw_ptr = buf->data_begin();
-    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
-    header->message_type = ReplicationMessageType::SHARD_MESSAGE;
-    header->payload_size = create_shard_message.size();
-    header->payload_crc =
-        crc32_ieee(init_crc32, r_cast< const uint8_t* >(create_shard_message.c_str()), create_shard_message.size());
-    header->header_crc = header->calculate_crc();
-    raw_ptr += sizeof(ReplicationMessageHeader);
-    std::memcpy(raw_ptr, create_shard_message.c_str(), create_shard_message.size());
 
+    // round up msg size to homestore datasvc blk size
+    const auto datasvc_blk_size = homestore::HomeStore::instance()->data_service().get_blk_size();
+    auto msg_size = create_shard_message.size();
+    if (msg_size % datasvc_blk_size != 0) { msg_size = (msg_size / datasvc_blk_size + 1) * datasvc_blk_size; }
+    auto msg_buf = iomanager.iobuf_alloc(512, msg_size);
+    std::memset(r_cast< uint8_t* >(msg_buf), 0, msg_size);
+    std::memcpy(r_cast< uint8_t* >(msg_buf), create_shard_message.c_str(), create_shard_message.size());
+
+    // preapre msg header;
+    ReplicationMessageHeader header;
+    header.repl_group_id = pg.value().pg_info.id;
+    header.msg_type = ReplicationMessageType::CREATE_SHARD_MSG;
+    header.payload_size = msg_size;
+    header.payload_crc = crc32_ieee(init_crc32, r_cast< const uint8_t* >(msg_buf), msg_size);
+    header.header_crc = header.calculate_crc();
     sisl::sg_list value;
+    value.size = msg_size;
+    value.iovs.push_back(iovec(msg_buf, msg_size));
     // replicate this create shard message to PG members;
     auto [p, sf] = folly::makePromiseContract< ShardManager::Result< ShardInfo > >();
-    replica_set->write(sisl::blob(buf->data_begin(), needed_size), sisl::blob(), value, static_cast< void* >(&p));
+    repl_dev.value()->async_alloc_write(sisl::blob(r_cast< uint8_t* >(&header), sizeof(header)), sisl::blob(), value,
+                                        static_cast< void* >(&p));
```

Comment:
I believe we also need to save the either the last allocated chunk_id assigned to this PG or the pdev_id hint itself to the context, so that when repldev calls get_alloc_blk_hints via listener, it knows what to return back as the hint, right?

Reply:
last allocated chunk_id is needed to pass down to HS, but not via ctx. it will be passed down by here: https://github.com/eBay/HomeObject/pull/68/files#diff-8f4c55993d40418d0dbb2db0f58c7d8696fd7d793f488104752d4b889f4efeaaR61.
I will complete this part since PR#70 is already merged now.